### ISSUE TITLE
judgmental equality with polymorphic universes

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#649]
Can we justify asking INRIA to fix this?:
```
Set Universe Polymorphism.
Inductive paths@{i} {A:Type@{i}} (a:A) : A -> Type@{i} := paths_refl : paths a a.
Inductive total2@{i} { T: Type@{i} } ( P: T -> Type@{i} ) : Type@{i} := tpair : forall (t:T) (p:P t), total2 P.
Set Printing Universes.
Set Printing All.
Section C.
  Universe i j.
  Constraint i < j.
  Lemma C (X:Type@{i}) (P:X -> Type@{i}) : paths@{j} (total2@{i} P) (total2@{j} P).
    reflexivity.                (* fails *)
```

**Comment by vladimirias** [comment]
We certainly can. However, as far as I know there is no other way to fix this one and similar issues, but by adopting the approach to the universe management that I have been suggesting since the summer of 2012. 

Mathieu Sozeau knows about it and the last time I heard him mention it, in Berlin, last summer, he said “we are still trying to do it our way”. 

After expressing my smug satisfaction with the fact that I am again confirmed to be right in what I was saying for a long time, I want to say that it will indeed be a good idea to remind them about this again and it would be great

**Comment by DanGrayson** [comment]
Good, I'll see what I can do.

Meanwhile, I'm working around by proving lots of little lemmas like this one:
```

  Lemma change_universe_pairs (X:Type@{i}) (P:X -> Type@{i}) :
    weq@{j} (total2@{i} P) (total2@{j} P).
  Proof.
    simple refine (weqpair _ (gradth _ _ _ _)).
    - intro w. try exact w. induction w as [x p]. exists x. exact p.
    - intro w. try exact w. induction w as [x p]. exists x. exact p.
    - intro w. try reflexivity. induction w as [x p]. reflexivity.
    - intro w. try reflexivity. induction w as [x p]. reflexivity.
  Defined.
```

**Comment by DanGrayson** [comment]
I've been working all weekend on dealing with universe polymorphism, and have encountered a choice that is forced on us by the problem I mentioned in the top message in this issue:

We have a choice:

   (1) we can make `unit` polymorphic, so that at each universe level i we have a unit
       type `unit@{i} : Type@{i} `, and then we have the problem that `unit@{i}`
       and `unit@{j}` are not the same unless i = j; or

   (2) we can make `unit` monomorphic, so that there is a single type
       `unit : Type@{u00}`, where u00 if a universe level constant.
       But then types like `paths tt tt` end up at universe
       level uu0, and `paths@{uu0} tt tt` is not the same as `paths@{i} tt tt` unless
       `uu0 = i`, so we have to be careful with the theorems that use paths in unit,
       for else they will force useless constraints of the form `uu0 = i`.  For example,
       in the theorem that says a type X equivalent to unit is contractible, X will be
       forced into universe level u00, which is unintended.

Maybe choice 1 is best if I want to continue experimenting for a while.

With either choice, if INRIA fixes Coq to add the judgmental equalities we want, we'll have
to remove some code.

So maybe it's best to wait to see if they're willing to fix that.  Or maybe I should submit a patch.

I think this issue is important, so we don't have to tiptoe around possible inconsistencies, as we did with the proof that a certain category has all limits, and so we can have much more confidence
in our formalizations.

**Comment by JasonGross** [comment]
See also https://sympa.inria.fr/sympa/arc/coq-club/2016-11/msg00017.html:

> Hi David,
>
> Yes, we're planning to work on it with Amin Timany for 8.7 or the following version.
> Do you have a particular application in mind?
>
> Best,
> -- Matthieu
>

**Comment by DanGrayson** [comment]
for Matthieu: The application we have in mind is to stop using type-in-type except for the two resizing axioms of Voevodsky.  Without doing that, we run a risk by using an system known to be unsound.

**Comment by JasonGross** [comment]
We might want to tag @mattam82, so he sees what you said about an application of cumulativity for inductives.

**Comment by vladimirias** [comment]
Importantly, it should not apply to types such as the equality type. It would be wrong if

a=_i b was convertible to a=_j b for i<j and a,b:Type@{i}

it should only apply to inductive definitions without indexes!!

**Comment by JasonGross** [comment]
> It would be wrong if
>
> a=_i b was convertible to a=_j b for i<j and a,b:Type@{i}

I don't think the proposal is to have even `total2@{i} P` and `total2@{j} P` be convertible; as I understand it, the proposal is to have `x : total2@{i} P` be typable as `x : total2@{j} P` whenever `i <= j`; similarly, to have `x : @paths@{i+1} Type@{i} a b` be typeable as `x : @paths@{j} Type@{i} a b` whenever `i < j`.

**Comment by DanGrayson** [comment]
Indeed, it was my proposal to have  `total2@{i} P` and `total2@{j} P` be convertible.
And we it implicitly includes convertibility of `@paths@{i} X x y` with `@paths@{i} X x y`.

Vladimir's objection, spelled out more carefully, would be that `@paths@{i} @Type@{i} X Y` and  `@paths@{j} @Type@{j} X Y` should not be convertible, and I agree with that.  But it's easy not to do that by mistake, because the two expressions differ by more than just a polymorphic universe level parameter.

**Comment by vladimirias** [comment]
I agree modulo the slight mistake we both made in the numbers. It should be paths@{k} Type@{i} X Y where k>i since we need Type@{i}:Type@{k} to apply paths@{k}.

**Comment by mattam82** [comment]
We're working on it with @amintimany right now. Currently, total2@{i} P becomes convertible with total2@{j} P (with P : A -> Type@{i} and i <= j) in our extended system. For @paths, it works similarly, @paths@{i} A x y = @paths@{j} A x y (A : Type@{i} and i <= j) but @paths@{i} Type@{j} x y <> @paths@{i} Type@{k} x y still, i.e convertibility is only possible if the parameters are themselves convertible already. Informally, two instances of I@{i} P and I@{j} P become convertible if one could write identity lifting functions between the two, e.g for paths:

```
  Constraint i < j.

  Lemma eq_up (A : Type@{i}) (x y : A) : paths@{i} x y -> paths@{j} x y.
  Proof.
    destruct 1. reflexivity.
  Qed.

  Lemma eq_down (A : Type@{i}) (x y : A) : paths@{j} x y -> paths@{i} x y.
  Proof.
    destruct 1. reflexivity.
  Qed.
```

**Comment by DanGrayson** [comment]
That sounds great, thanks, Matthieu!  (@mattam82) Can we expect it for 8.7?

I assume that instead of "two instances of I@{i} P and I@{j} P become convertible" you meant "I@{i} P and I@{j} P become convertible".

**Comment by DanGrayson** [comment]
PS: @mattam82 @amintimany -- I'd be happy (and eager) to be an α-tester for your code, once it starts doing something.

**Comment by mattam82** [comment]
@DanGrayson yes that's right. We're looking at the implementation now, I'll keep you updated when we have something working!
